CdnEngine_Azure_MI.php

```diff
@@ -0,0 +1,414 @@
+		}
+
+		$contents = @file_get_contents( $local_path );
+		$md5 = md5( $contents );   // @md5_file( $local_path );
+		$content_md5 = $this->_get_content_md5( $md5 );
+
+		if ( !$force_rewrite ) {
+			try {
+				$p = CdnEngine_Azure_MI_Utility::getBlobProperties( $this->_config['clientid'],
+					$this->_config['user'], $this->_config['container'], $remote_path );
+				$local_size = @filesize( $local_path );
+
+				//check if Content-Length is available in $p array
+				if ( isset( $p['Content-Length']) && $local_size == $p['Content-Length'] 
+					&& isset( $p['Content-MD5']) && $content_md5 === $p['Content-MD5'] ) {
+					return $this->_get_result( $local_path, $remote_path,
+						W3TC_CDN_RESULT_OK, 'File up-to-date.', $file );
+				}
+			} catch ( \Exception $exception ) {
+			}
+		}
+
+		$headers = $this->get_headers_for_file( $file );
+
+		try {
+			$contentType = isset($headers['Content-Type']) ? $headers['Content-Type'] : 'application/octet-stream';
+			$cacheControl = isset($headers['Cache-Control']) ? $headers['Cache-Control'] : '';
+
+			CdnEngine_Azure_MI_Utility::createBlockBlob( $this->_config['clientid'], $this->_config['user'], 
+				$this->_config['container'], $remote_path, $contents, $contentType, $content_md5, $cacheControl);
+			
+		} catch ( \Exception $exception ) {
+			return $this->_get_result( $local_path, $remote_path,
+				W3TC_CDN_RESULT_ERROR,
+				sprintf( 'Unable to put blob (%s).', $exception->getMessage() ),
+				$file );
+		}
+
+		return $this->_get_result( $local_path, $remote_path, W3TC_CDN_RESULT_OK,
+			'OK', $file );
+	}
+
+	/**
+	 * Deletes files from storage
+	 *
+	 * @param array   $files
+	 * @param array   $results
+	 * @return boolean
+	 */
+	function delete( $files, &$results ) {
+		$error = null;
+
+		if ( !$this->_init( $error ) ) {
+			$results = $this->_get_results( $files, W3TC_CDN_RESULT_HALT, $error );
+
+			return false;
+		}
+
+		foreach ( $files as $file ) {
+			$local_path = $file['local_path'];
+			$remote_path = $file['remote_path'];
+
+			try {
+
+				CdnEngine_Azure_MI_Utility::deleteBlob( $this->_config['clientid'], $this->_config['user'], 
+					$this->_config['container'], $remote_path );
+				$results[] = $this->_get_result( $local_path, $remote_path,
+					W3TC_CDN_RESULT_OK, 'OK', $file );
+			} catch ( \Exception $exception ) {
+				$results[] = $this->_get_result( $local_path, $remote_path,
+					W3TC_CDN_RESULT_ERROR,
+					sprintf( 'Unable to delete blob (%s).', $exception->getMessage() ),
+					$file );
+			}
+		}
+
+		return !$this->_is_error( $results );
+	}
+
+	/**
+	 * Tests S3
+	 *
+	 * @param string  $error
+	 * @return boolean
+	 */
+	function test( &$error ) {
+		if ( !parent::test( $error ) ) {
+			return false;
+		}
+
+		$string = 'test_azure_' . md5( time() );
+
+		if ( !$this->_init( $error ) ) {
+			return false;
+		}
+
+		try {
+			$containers = CdnEngine_Azure_MI_Utility::listContainers( $this->_config['clientid'],
```

Comment:
You can use single container GET API instead of LIST API here 
https://learn.microsoft.com/en-us/rest/api/storageservices/get-container-properties?tabs=microsoft-entra-id

Reply:
We can probably take this up in next versions. This is anyways only for 'Test' button functionality and not for actual operations. 

The reason is that I am trying to keep code changes to minimum and just replacing the calls to blob storage functions while keep the rest of the code and core logic same as it was before.